// @dapr/dapr/workflow/client/WorkflowFailureDetails.js
"use strict";
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowFailureDetails = void 0;
/**
 * Class that represents the details of a task failure.
 *
 * In most cases, failures are caused by unhandled exceptions in activity or workflow code, in which case instances
 * of this class will expose the details of the exception. However, it's also possible that other types of errors could
 * result in task failures, in which case there may not be any exception-specific information.
 */
class WorkflowFailureDetails {
    constructor(failureDetails) {
        this.failureDetails = failureDetails;
    }
    /**
     * Gets the error type, which is the namespace-qualified exception type name.
     * @return {string} The error type.
     */
    getErrorType() {
        return this.failureDetails.errorType;
    }
    /**
     * Gets the error message.
     * @return {string} The error message.
     */
    getErrorMessage() {
        return this.failureDetails.message;
    }
    /**
     * Gets the stack trace.
     * @return {string | undefined} The stack trace, or undefined if not available.
     */
    getStackTrace() {
        return this.failureDetails.stackTrace;
    }
}
exports.WorkflowFailureDetails = WorkflowFailureDetails;
