// @dapr/dapr/workflow/client/WorkflowState.js
"use strict";
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowState = void 0;
const WorkflowFailureDetails_1 = require("./WorkflowFailureDetails");
const WorkflowRuntimeStatus_1 = require("../runtime/WorkflowRuntimeStatus");
/**
 * Represents the state of a workflow instance.
 */
class WorkflowState {
    /**
     * Creates an instance of WorkflowState.
     * @param {OrchestrationState} orchestrationState - The state of the orchestration.
     * @throws {Error} Throws an error if orchestrationState is null.
     */
    constructor(orchestrationState) {
        if (!orchestrationState) {
            throw new Error("OrchestrationMetadata cannot be null");
        }
        this._orchestrationState = orchestrationState;
        const failureDetails = orchestrationState.failureDetails;
        if (failureDetails) {
            this._workflowFailureDetails = new WorkflowFailureDetails_1.WorkflowFailureDetails(failureDetails);
        }
    }
    /**
     * Gets the name of the workflow.
     * @returns {string} The name of the workflow.
     */
    get name() {
        return this._orchestrationState.name;
    }
    /**
     * Gets the unique ID of the workflow instance.
     * @returns {string} The unique ID of the workflow instance.
     */
    get instanceId() {
        return this._orchestrationState.instanceId;
    }
    /**
     * Gets the current runtime status of the workflow instance.
     * @returns {WorkflowRuntimeStatus} The current runtime status.
     */
    get runtimeStatus() {
        return (0, WorkflowRuntimeStatus_1.fromOrchestrationStatus)(this._orchestrationState.runtimeStatus);
    }
    /**
     * Gets the workflow instance's creation time in UTC.
     * @returns {Date} The workflow instance's creation time in UTC.
     */
    get createdAt() {
        return this._orchestrationState.createdAt;
    }
    /**
     * Gets the workflow instance's last updated time in UTC.
     * @returns {Date} The workflow instance's last updated time in UTC.
     */
    get lastUpdatedAt() {
        return this._orchestrationState.lastUpdatedAt;
    }
    /**
     * Gets the workflow instance's serialized input, if any, as a string value.
     * @returns {string | undefined} The workflow instance's serialized input or undefined.
     */
    get serializedInput() {
        return this._orchestrationState.serializedInput;
    }
    /**
     * Gets the workflow instance's serialized output, if any, as a string value.
     * @returns {string | undefined} The workflow instance's serialized output or undefined.
     */
    get serializedOutput() {
        return this._orchestrationState.serializedOutput;
    }
    /**
     * Gets the failure details, if any, for the failed workflow instance.
     * This method returns data only if the workflow is in the FAILED state and
     * only if this instance metadata was fetched with the option to include output data.
     * @returns {WorkflowFailureDetails | undefined} The failure details of the failed workflow instance or undefined.
     */
    get workflowFailureDetails() {
        return this._workflowFailureDetails;
    }
}
exports.WorkflowState = WorkflowState;
