// @dapr/dapr/implementation/Client/GRPCClient/configuration.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const dapr_pb_1 = require("../../../proto/dapr/proto/runtime/v1/dapr_pb");
const Client_util_1 = require("../../../utils/Client.util");
class GRPCClientConfiguration {
    constructor(client) {
        this.client = client;
    }
    async get(storeName, keys, metadataObj) {
        const msg = new dapr_pb_1.GetConfigurationRequest();
        msg.setStoreName(storeName);
        if (keys && keys.length > 0) {
            msg.setKeysList(keys.filter((i) => i !== ""));
        }
        (0, Client_util_1.addMetadataToMap)(msg.getMetadataMap(), metadataObj);
        const client = await this.client.getClient();
        return new Promise((resolve, reject) => {
            client.getConfiguration(msg, (err, res) => {
                if (err) {
                    return reject(err);
                }
                const configMap = (0, Client_util_1.createConfigurationType)(res.getItemsMap());
                const result = {
                    items: configMap,
                };
                return resolve(result);
            });
        });
    }
    async subscribe(storeName, cb) {
        return this._subscribe(storeName, cb);
    }
    async subscribeWithKeys(storeName, keys, cb) {
        return this._subscribe(storeName, cb, keys);
    }
    async subscribeWithMetadata(storeName, keys, metadata, cb) {
        return this._subscribe(storeName, cb, keys, metadata);
    }
    async _subscribe(storeName, cb, keys, metadataObj) {
        const msg = new dapr_pb_1.SubscribeConfigurationRequest();
        msg.setStoreName(storeName);
        if (keys && keys.length > 0) {
            msg.setKeysList(keys.filter((i) => i !== ""));
        }
        else {
            msg.setKeysList([]);
        }
        (0, Client_util_1.addMetadataToMap)(msg.getMetadataMap(), metadataObj);
        const client = await this.client.getClient();
        // Open a stream. Note that this is a never-ending stream
        // and will stay open as long as the client is open
        // we will thus create a set with our listeners so we don't
        // break on multi listeners
        const stream = client.subscribeConfiguration(msg);
        let streamId;
        stream.on("data", async (data) => {
            streamId = data.getId();
            const items = data.getItemsMap();
            if (items.getLength() == 0) {
                return;
            }
            const configMap = (0, Client_util_1.createConfigurationType)(items);
            const wrapped = {
                items: configMap,
            };
            await cb(wrapped);
        });
        return {
            stop: async () => {
                return new Promise((resolve, reject) => {
                    const req = new dapr_pb_1.UnsubscribeConfigurationRequest();
                    req.setStoreName(storeName);
                    req.setId(streamId);
                    client.unsubscribeConfiguration(req, (err, res) => {
                        if (err || !res.getOk()) {
                            return reject(res.getMessage());
                        }
                        // Clean up the node.js event emitter
                        stream.removeAllListeners();
                        stream.destroy();
                        return resolve();
                    });
                });
            },
        };
    }
}
exports.default = GRPCClientConfiguration;
