// @dapr/dapr/actors/runtime/ActorReminderData.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BufferSerializer_1 = __importDefault(require("./BufferSerializer"));
/**
 * Contains the actor reminder data
 */
class ActorReminderData {
    /**
     * @param reminderName the name of the actor reminder
     * @param state the state data passed to receiveReminder callback
     * @param dueTime the amount of time to delay before invoking the reminder for the first time
     * @param ttl time to duration after which the reminder will be expired and deleted
     * @param period the time interval between reminder invocations after the first invocation
     */
    constructor(reminderName, dueTime, period, ttl, state) {
        this.reminderName = reminderName;
        this.dueTime = dueTime;
        this.ttl = ttl;
        this.period = period;
        this.state = state;
    }
    getReminderName() {
        return this.reminderName;
    }
    getState() {
        return this.state;
    }
    getDueTime() {
        return this.dueTime;
    }
    getTtl() {
        return this.ttl;
    }
    getPeriod() {
        return this.period;
    }
    /**
     * Return this class as an object
     */
    toObject() {
        return {
            reminderName: this.reminderName,
            dueTime: this.dueTime,
            ttl: this.ttl,
            period: this.period,
            data: this.state,
        };
    }
    static fromObject(reminderName, obj) {
        const serializer = new BufferSerializer_1.default();
        const data = obj?.data;
        const dueTime = obj?.dueTime;
        const ttl = obj?.ttl;
        const period = obj?.period;
        const deserializedData = serializer.deserialize(data);
        return new ActorReminderData(reminderName, dueTime, ttl, period, deserializedData);
    }
}
exports.default = ActorReminderData;
