// @dapr/dapr/implementation/Client/HTTPClient/lock.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const UnlockResponse_1 = require("../../../types/lock/UnlockResponse");
class HTTPClientLock {
    constructor(client) {
        this.client = client;
    }
    async lock(storeName, resourceId, lockOwner, expiryInSeconds) {
        const data = {
            resourceId,
            lockOwner,
            expiryInSeconds,
        };
        const result = await this.client.executeWithApiVersion("v1.0-alpha1", `/lock/${storeName}`, {
            method: "POST",
            body: data,
        });
        return {
            success: result["success"],
        };
    }
    async unlock(storeName, resourceId, lockOwner) {
        const data = {
            resourceId,
            lockOwner,
        };
        const result = await this.client.executeWithApiVersion("v1.0-alpha1", `/unlock/${storeName}`, {
            method: "POST",
            body: data,
        });
        return {
            status: this._statusToLockStatus(result["status"]),
        };
    }
    _statusToLockStatus(status) {
        switch (status) {
            case 0:
                return UnlockResponse_1.LockStatus.Success;
            case 1:
                return UnlockResponse_1.LockStatus.LockDoesNotExist;
            case 2:
                return UnlockResponse_1.LockStatus.LockBelongsToOthers;
            case 3:
            default:
                return UnlockResponse_1.LockStatus.InternalError;
        }
    }
}
exports.default = HTTPClientLock;
