// @dapr/dapr/implementation/Client/HTTPClient/workflow.js
"use strict";
/*
Copyright 2023 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const Logger_1 = require("../../../logger/Logger");
const crypto_1 = require("crypto");
const Client_util_1 = require("../../../utils/Client.util");
const PropertyRequiredError_1 = require("../../../errors/PropertyRequiredError");
class HTTPClientWorkflow {
    constructor(client) {
        this.client = client;
        this.logger = new Logger_1.Logger("HTTPClient", "Workflow", client.options.logger);
    }
    async get(instanceID, workflowComponent) {
        if (!instanceID) {
            throw new PropertyRequiredError_1.PropertyRequiredError("instanceID");
        }
        workflowComponent = workflowComponent ?? HTTPClientWorkflow.DEFAULT_WORKFLOW_COMPONENT;
        try {
            const result = await this.client.executeWithApiVersion("v1.0-beta1", `/workflows/${workflowComponent}/${instanceID}`, { method: "GET" });
            if (typeof result === "string") {
                throw new Error(`Error getting workflow instance: ${result}`);
            }
            const resultJson = result;
            const response = {
                instanceID: resultJson["instanceID"],
                workflowName: resultJson["workflowName"],
                createdAt: resultJson["createdAt"] ? new Date(resultJson["createdAt"]) : new Date(),
                lastUpdatedAt: resultJson["lastUpdatedAt"] ? new Date(resultJson["lastUpdatedAt"]) : new Date(),
                runtimeStatus: resultJson["runtimeStatus"],
                properties: resultJson["properties"],
            };
            return response;
        }
        catch (e) {
            this.logger.error(`Error getting workflow instance: ${e.message}`);
            throw e;
        }
    }
    async start(workflowName, input, instanceId, workflowComponent, options = {}) {
        if (!workflowName) {
            throw new PropertyRequiredError_1.PropertyRequiredError("workflowName");
        }
        if (!instanceId) {
            instanceId = (0, crypto_1.randomUUID)();
        }
        workflowComponent = workflowComponent ?? HTTPClientWorkflow.DEFAULT_WORKFLOW_COMPONENT;
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: { instanceID: instanceId } });
        // Set content type if provided.
        // If not, HTTPClient will infer it from the data.
        const headers = {};
        if (options.contentType) {
            headers["Content-Type"] = options.contentType;
        }
        try {
            await this.client.executeWithApiVersion("v1.0-beta1", `/workflows/${workflowComponent}/${workflowName}/start?${queryParams}`, {
                method: "POST",
                body: input,
                headers,
            });
            return instanceId;
        }
        catch (e) {
            this.logger.error(`Error starting workflow instance: ${e.message}`);
            throw e;
        }
    }
    async raise(instanceId, eventName, eventData, workflowComponent, options = {}) {
        if (!instanceId) {
            throw new PropertyRequiredError_1.PropertyRequiredError("instanceID");
        }
        if (!eventName) {
            throw new PropertyRequiredError_1.PropertyRequiredError("eventName");
        }
        workflowComponent = workflowComponent ?? HTTPClientWorkflow.DEFAULT_WORKFLOW_COMPONENT;
        // Set content type if provided.
        // If not, HTTPClient will infer it from the data.
        const headers = {};
        if (options.eventContentType) {
            headers["Content-Type"] = options.eventContentType;
        }
        try {
            await this.client.executeWithApiVersion("v1.0-beta1", `/workflows/${workflowComponent}/${instanceId}/raiseEvent/${eventName}`, {
                method: "POST",
                body: eventData,
                headers,
            });
        }
        catch (e) {
            this.logger.error(`Error raising event on workflow instance: ${e.message}`);
            throw e;
        }
    }
    async terminate(instanceId, workflowComponent) {
        await this._invokeMethod(instanceId, "terminate", workflowComponent);
    }
    async pause(instanceId, workflowComponent) {
        await this._invokeMethod(instanceId, "pause", workflowComponent);
    }
    async resume(instanceId, workflowComponent) {
        await this._invokeMethod(instanceId, "resume", workflowComponent);
    }
    async purge(instanceId, workflowComponent) {
        await this._invokeMethod(instanceId, "purge", workflowComponent);
    }
    async _invokeMethod(instanceId, method, workflowComponent) {
        if (!instanceId) {
            throw new PropertyRequiredError_1.PropertyRequiredError("instanceID");
        }
        if (!method) {
            throw new PropertyRequiredError_1.PropertyRequiredError("method");
        }
        workflowComponent = workflowComponent ?? HTTPClientWorkflow.DEFAULT_WORKFLOW_COMPONENT;
        try {
            await this.client.executeWithApiVersion("v1.0-beta1", `/workflows/${workflowComponent}/${instanceId}/${method}`, {
                method: "POST",
            });
        }
        catch (e) {
            this.logger.error(`Error invoking ${method} on workflow instance: ${e.message}`);
            throw e;
        }
    }
}
exports.default = HTTPClientWorkflow;
HTTPClientWorkflow.DEFAULT_WORKFLOW_COMPONENT = "dapr";
