// @dapr/dapr/implementation/Client/HTTPClient/pubsub.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const Logger_1 = require("../../../logger/Logger");
const Client_util_1 = require("../../../utils/Client.util");
// https://docs.dapr.io/reference/api/pubsub_api/
class HTTPClientPubSub {
    constructor(client) {
        this.client = client;
        this.logger = new Logger_1.Logger("HTTPClient", "PubSub", client.options.logger);
    }
    async publish(pubSubName, topic, data, options = {}) {
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: options.metadata, type: "metadata" });
        // Set content type if provided.
        // If not, HTTPClient will infer it from the data.
        const headers = {};
        if (options.contentType) {
            headers["Content-Type"] = options.contentType;
        }
        try {
            await this.client.execute(`/publish/${pubSubName}/${topic}?${queryParams}`, {
                method: "POST",
                body: data,
                headers,
            });
        }
        catch (e) {
            this.logger.error(`publish failed: ${e}`);
            return { error: e };
        }
        return {};
    }
    async publishBulk(pubSubName, topic, messages, metadata) {
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: metadata, type: "metadata" });
        const params = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
        };
        const entries = (0, Client_util_1.getBulkPublishEntries)(messages);
        params.body = entries;
        try {
            await this.client.executeWithApiVersion("v1.0-alpha1", `/publish/bulk/${pubSubName}/${topic}?${queryParams}`, params);
        }
        catch (error) {
            this.logger.error(`Failure publishing bulk messages: ${error}`);
            return this.handleBulkPublishError(entries, error);
        }
        // If no error is thrown, all messages were published successfully
        return { failedMessages: [] };
    }
    async handleBulkPublishError(entries, error) {
        try {
            // If the error is returned by the bulk publish API,
            // parse the error message and return the response
            const err = JSON.parse(error.message);
            if (err.error_msg) {
                const bulkPublishResponse = JSON.parse(err.error_msg);
                return (0, Client_util_1.getBulkPublishResponse)({ entries: entries, response: bulkPublishResponse });
            }
        }
        catch (_innerError) {
            // This can indicate a general error with the request (e.g., network error, invalid pubsub name, etc.).
        }
        return (0, Client_util_1.getBulkPublishResponse)({ entries: entries, error: error });
    }
}
exports.default = HTTPClientPubSub;
