// @dapr/dapr/implementation/Client/HTTPClient/state.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const Client_util_1 = require("../../../utils/Client.util");
const Settings_util_1 = require("../../../utils/Settings.util");
const Logger_1 = require("../../../logger/Logger");
// https://docs.dapr.io/reference/api/state_api/
class HTTPClientState {
    constructor(client) {
        this.client = client;
        this.logger = new Logger_1.Logger("HTTPClient", "State", client.options.logger);
    }
    async save(storeName, stateObjects, options = {}) {
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: options?.metadata, type: "metadata" });
        for (const so of stateObjects) {
            const behavior = {
                consistency: (0, Client_util_1.getStateConsistencyValue)(so?.options?.consistency),
                concurrency: (0, Client_util_1.getStateConcurrencyValue)(so?.options?.concurrency),
            };
            so.options = Object.assign({}, so.options, behavior);
        }
        try {
            await this.client.execute(`/state/${storeName}?${queryParams}`, {
                method: "POST",
                body: stateObjects,
            });
        }
        catch (e) {
            this.logger.error(`Error saving state to store ${storeName}, error: ${e}`);
            return { error: e };
        }
        return {};
    }
    async get(storeName, key, options) {
        const behavior = {
            consistency: (0, Client_util_1.getStateConsistencyValue)(options?.consistency),
        };
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: options?.metadata, type: "metadata" }, { data: behavior });
        const result = await this.client.execute(`/state/${storeName}/${key}?${queryParams}`);
        return result;
    }
    async getBulk(storeName, keys, options) {
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: options?.metadata, type: "metadata" });
        const result = await this.client.execute(`/state/${storeName}/bulk?${queryParams}`, {
            method: "POST",
            body: {
                keys,
                parallelism: options?.parallelism ?? Settings_util_1.Settings.getDefaultStateGetBulkParallelism,
            },
        });
        return result;
    }
    async delete(storeName, key, options) {
        const behavior = {
            concurrency: (0, Client_util_1.getStateConcurrencyValue)(options?.concurrency),
            consistency: (0, Client_util_1.getStateConsistencyValue)(options?.consistency),
        };
        const queryParams = (0, Client_util_1.createHTTPQueryParam)({ data: options?.metadata, type: "metadata" }, { data: behavior });
        // Managed headers
        const headers = {};
        if (options?.etag) {
            headers["If-Match"] = options.etag;
        }
        try {
            await this.client.execute(`/state/${storeName}/${key}?${queryParams}`, {
                method: "DELETE",
                headers,
            });
        }
        catch (e) {
            this.logger.error(`Error deleting state from store ${storeName}, error: ${e}`);
            return { error: e };
        }
        return {};
    }
    async transaction(storeName, operations = [], metadata = null) {
        for (const op of operations) {
            const behavior = {
                consistency: (0, Client_util_1.getStateConsistencyValue)(op?.request?.options?.consistency),
                concurrency: (0, Client_util_1.getStateConcurrencyValue)(op?.request.options?.concurrency),
            };
            op.request.options = Object.assign({}, op.request.options, behavior);
        }
        await this.client.execute(`/state/${storeName}/transaction`, {
            method: "POST",
            body: {
                operations,
                metadata,
            },
        });
    }
    async query(storeName, query) {
        const result = await this.client.executeWithApiVersion("v1.0-alpha1", `/state/${storeName}/query`, {
            method: "POST",
            body: {
                ...query,
            },
        });
        if (result === "") {
            return { results: [] };
        }
        return result;
    }
}
exports.default = HTTPClientState;
