// @microsoft/durabletask-js/orchestration/orchestration-purge-criteria.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PurgeInstanceCriteria = void 0;
class PurgeInstanceCriteria {
    constructor() {
        this.runtimeStatusList = [];
        // default timeout 5 mins
        this.timeout = 5 * 60 * 1000;
    }
    // Setter methods to allow users to set values later
    setCreatedTimeFrom(date) {
        this.createdTimeFrom = date;
    }
    getCreatedTimeFrom() {
        return this.createdTimeFrom;
    }
    setCreatedTimeTo(date) {
        this.createdTimeTo = date;
    }
    getCreatedTimeTo() {
        return this.createdTimeTo;
    }
    setRuntimeStatusList(statusList) {
        this.runtimeStatusList = statusList;
    }
    getRuntimeStatusList() {
        return this.runtimeStatusList;
    }
    setTimeout(timeout) {
        this.timeout = timeout;
    }
    getTimeout() {
        return this.timeout;
    }
}
exports.PurgeInstanceCriteria = PurgeInstanceCriteria;
