// @dapr/dapr/implementation/Client/GRPCClient/GRPCClientProxy.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GRPCClientProxy = void 0;
const grpc = __importStar(require("@grpc/grpc-js"));
const Settings_util_1 = require("../../../utils/Settings.util");
class GRPCClientProxy {
    constructor(clsProxy, grpcClient, grpcClientOptions) {
        this.clsProxy = clsProxy;
        this.grpcClient = grpcClient;
        this.grpcClientOptions = grpcClientOptions ?? {};
    }
    generateInterceptors() {
        const interceptors = [];
        // Create an interceptor that adds 'dapr-app-id' to each call as metadata
        const interceptorDaprAppId = (options, nextCall) => {
            return new grpc.InterceptingCall(nextCall(options), {
                start: (metadata, listener, next) => {
                    metadata.add("dapr-app-id", `${Settings_util_1.Settings.getAppId()}`);
                    next(metadata, listener);
                },
            });
        };
        interceptors.push(interceptorDaprAppId);
        return interceptors;
    }
    async build() {
        // Ensures the sidecar is started
        await this.grpcClient.start();
        // Call our method as defined in the proto
        if (!this.grpcClientOptions.interceptors) {
            this.grpcClientOptions.interceptors = [];
        }
        this.grpcClientOptions.interceptors = [
            ...this.generateInterceptors(),
            ...(this.grpcClient.getGrpcClientOptions().interceptors ?? []),
            ...this.grpcClientOptions.interceptors,
        ];
        const clientCustom = new this.clsProxy(`${this.grpcClient.options.daprHost}:${this.grpcClient.options.daprPort}`, this.grpcClient.getClientCredentials(), this.grpcClientOptions);
        return clientCustom;
    }
}
exports.GRPCClientProxy = GRPCClientProxy;
