// @dapr/dapr/implementation/Client/GRPCClient/workflow.js
"use strict";
/*
Copyright 2023 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const GRPCNotSupportedError_1 = require("../../../errors/GRPCNotSupportedError");
class GRPCClientWorkflow {
    constructor(client) {
        this.client = client;
    }
    get(_instanceId, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
    start(_workflowName, _input, _instanceId, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
    terminate(_instanceId, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
    pause(_instanceId, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
    resume(_instanceId, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
    purge(_instanceId, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
    raise(_instanceId, _eventName, _input, _workflowComponent) {
        throw new GRPCNotSupportedError_1.GRPCNotSupportedError();
    }
}
exports.default = GRPCClientWorkflow;
