// @dapr/dapr/workflow/internal/index.js
"use strict";
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateApiTokenClientInterceptors = exports.getDaprApiToken = exports.generateEndpoint = exports.getFunctionName = void 0;
const grpc = __importStar(require("@grpc/grpc-js"));
const Settings_util_1 = require("../../utils/Settings.util");
const GrpcEndpoint_1 = require("../../network/GrpcEndpoint");
/**
 * Gets the name of a function from its definition or string representation.
 *
 * @param fn - The function for which the name is to be retrieved. Can be either a function or a string representation of a function.
 * @returns The name of the function.
 *
 * @typeparam TWorkflow - The type of the workflow function.
 * @typeparam TInput - The type of the input for the workflow activity.
 * @typeparam TOutput - The type of the output for the workflow activity.
 */
function getFunctionName(fn) {
    if (fn.name) {
        return fn.name;
    }
    else {
        const match = fn.toString().match(/function\s*([^(]*)\(/);
        if (match === null) {
            throw new Error("Unable to determine function name, try to sepecify the workflow/activity name explicitly.");
        }
        return match[1];
    }
}
exports.getFunctionName = getFunctionName;
/**
 * Generates a Dapr gRPC endpoint based on the provided options or defaults.
 *
 * @param options - An object containing optional settings for the WorkflowClient.
 * @returns A GrpcEndpoint representing the Dapr gRPC endpoint.
 */
function generateEndpoint(options) {
    // NOTE: same logic as src/implementation/Client/GRPCClient/GRPCClient.ts
    const host = options?.daprHost ?? Settings_util_1.Settings.getDefaultHost();
    const port = options?.daprPort ?? Settings_util_1.Settings.getDefaultGrpcPort();
    let uri = `${host}:${port}`;
    if (!(options?.daprHost || options?.daprPort)) {
        // If neither host nor port are specified, check the endpoint environment variable.
        const endpoint = Settings_util_1.Settings.getDefaultGrpcEndpoint();
        if (endpoint != "") {
            uri = endpoint;
        }
    }
    return new GrpcEndpoint_1.GrpcEndpoint(uri);
}
exports.generateEndpoint = generateEndpoint;
/**
 * Gets the Dapr API token based on the provided options or defaults.
 *
 * @param options - An object containing optional settings for the WorkflowClient.
 * @returns A string representing the Dapr API token, or undefined if not set.
 */
function getDaprApiToken(options) {
    const daprApiToken = options?.daprApiToken ?? Settings_util_1.Settings.getDefaultApiToken();
    return daprApiToken;
}
exports.getDaprApiToken = getDaprApiToken;
/**
 * Generates a gRPC interceptor function that adds a Dapr API token to the metadata if not already present.
 * This interceptor is intended for use with gRPC client calls.
 *
 * @param options - The gRPC call options object.
 * @param nextCall - The next call function in the interceptor chain.
 * @returns A gRPC InterceptingCall instance with added functionality to include a Dapr API token in the metadata.
 */
function generateApiTokenClientInterceptors(workflowOptions) {
    return (options, nextCall) => {
        return new grpc.InterceptingCall(nextCall(options), {
            start: (metadata, listener, next) => {
                if (metadata.get("dapr-api-token").length == 0) {
                    metadata.add("dapr-api-token", workflowOptions.daprApiToken);
                }
                next(metadata, listener);
            },
        });
    };
}
exports.generateApiTokenClientInterceptors = generateApiTokenClientInterceptors;
