// @dapr/dapr/workflow/runtime/WorkflowRuntimeStatus.js
"use strict";
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.toOrchestrationStatus = exports.fromOrchestrationStatus = exports.WorkflowRuntimeStatus = void 0;
const orchestration_status_enum_1 = require("@microsoft/durabletask-js/orchestration/enum/orchestration-status.enum");
/**
 * Enum describing the runtime status of a workflow.
 */
var WorkflowRuntimeStatus;
(function (WorkflowRuntimeStatus) {
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["RUNNING"] = 0] = "RUNNING";
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["COMPLETED"] = 1] = "COMPLETED";
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["FAILED"] = 3] = "FAILED";
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["TERMINATED"] = 5] = "TERMINATED";
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["CONTINUED_AS_NEW"] = 2] = "CONTINUED_AS_NEW";
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["PENDING"] = 6] = "PENDING";
    WorkflowRuntimeStatus[WorkflowRuntimeStatus["SUSPENDED"] = 7] = "SUSPENDED";
})(WorkflowRuntimeStatus = exports.WorkflowRuntimeStatus || (exports.WorkflowRuntimeStatus = {}));
/**
 * Converts an OrchestrationStatus value to the corresponding WorkflowRuntimeStatus enum value.
 *
 * @param {OrchestrationStatus} val - The OrchestrationStatus value to be converted.
 * @returns {WorkflowRuntimeStatus} - The equivalent WorkflowRuntimeStatus enum value.
 */
function fromOrchestrationStatus(val) {
    const values = Object.values(WorkflowRuntimeStatus);
    const valIdx = values.findIndex((v) => v == val);
    // Return the entry of the WorkflowRuntimeStatus enum at index
    const entries = Object.entries(WorkflowRuntimeStatus);
    return entries[valIdx][1];
}
exports.fromOrchestrationStatus = fromOrchestrationStatus;
/**
 * Converts an WorkflowRuntimeStatus value to the corresponding OrchestrationStatus enum value.
 *
 * @param {WorkflowRuntimeStatus} val - The WorkflowRuntimeStatus value to be converted.
 * @returns {OrchestrationStatus} - The equivalent OrchestrationStatus enum value.
 */
function toOrchestrationStatus(val) {
    const values = Object.values(orchestration_status_enum_1.OrchestrationStatus);
    const valIdx = values.findIndex((v) => v == val);
    // Return the entry of the WorkflowRuntimeStatus enum at index
    const entries = Object.entries(orchestration_status_enum_1.OrchestrationStatus);
    return entries[valIdx][1];
}
exports.toOrchestrationStatus = toOrchestrationStatus;
