// @dapr/dapr/network/AbstractEndpoint.js
"use strict";
/*
Copyright 2023 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoint = void 0;
class Endpoint {
    constructor(url) {
        this._scheme = "";
        this._hostname = "";
        this._port = 0;
        this._tls = false;
        this._endpoint = "";
        this._url = url;
    }
    get tls() {
        return this._tls;
    }
    get hostname() {
        return this._hostname;
    }
    get scheme() {
        return this._scheme;
    }
    get port() {
        return this._port === 0 ? "" : this._port.toString();
    }
    get endpoint() {
        return this._endpoint;
    }
}
exports.Endpoint = Endpoint;
