// @dapr/dapr/actors/client/ActorClient/ActorClientHTTP.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
// https://docs.dapr.io/reference/api/actors_api/
class ActorClientHTTP {
    constructor(client) {
        this.client = client;
    }
    async invoke(actorType, actorId, methodName, body) {
        const result = await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/method/${methodName}`, {
            method: "POST",
            body,
        });
        return result;
    }
    async stateTransaction(actorType, actorId, operations) {
        await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/state`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: operations,
        });
    }
    async stateGet(actorType, actorId, key) {
        const result = await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/state/${key}`);
        return result;
    }
    async registerActorReminder(actorType, actorId, name, reminder) {
        await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/reminders/${name}`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: {
                period: reminder?.period?.toString()?.toLocaleLowerCase().replace("pt", ""),
                dueTime: reminder?.dueTime?.toString()?.toLocaleLowerCase().replace("pt", ""),
                ttl: reminder?.ttl?.toString()?.toLocaleLowerCase().replace("pt", ""),
                data: reminder.data,
            },
        });
    }
    async reminderGet(actorType, actorId, name) {
        const result = await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/reminders/${name}`);
        return result;
    }
    async unregisterActorReminder(actorType, actorId, name) {
        await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/reminders/${name}`, {
            method: "DELETE",
        });
    }
    async registerActorTimer(actorType, actorId, name, timer) {
        await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/timers/${name}`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: {
                period: timer?.period?.toString()?.toLocaleLowerCase().replace("pt", ""),
                dueTime: timer?.dueTime?.toString()?.toLocaleLowerCase().replace("pt", ""),
                ttl: timer?.ttl?.toString()?.toLocaleLowerCase().replace("pt", ""),
                data: timer.data,
                callback: timer.callback,
            },
        });
    }
    async unregisterActorTimer(actorType, actorId, name) {
        await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}/timers/${name}`, {
            method: "DELETE",
        });
    }
    async deactivate(actorType, actorId) {
        await this.client.execute(`/actors/${actorType}/${actorId.getURLSafeId()}`, {
            method: "DELETE",
        });
    }
    async getActors() {
        const result = await this.client.execute(`replace('/v1.0', '')}/dapr/config`);
        return result;
    }
}
exports.default = ActorClientHTTP;
