// @dapr/dapr/actors/runtime/ActorMethodContext.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const ActorCallType_1 = require("./ActorCallType");
class ActorMethodContext {
    constructor(methodName, callType) {
        this.methodName = methodName;
        this.callType = callType;
    }
    getMethodName() {
        return this.methodName;
    }
    getCallType() {
        return this.callType;
    }
    static createForActor(methodName) {
        return new ActorMethodContext(methodName, ActorCallType_1.ActorCallType.ACTOR_INTERFACE_METHOD);
    }
    static createForTimer(methodName) {
        return new ActorMethodContext(methodName, ActorCallType_1.ActorCallType.TIMER_METHOD);
    }
    static createForReminder(methodName) {
        return new ActorMethodContext(methodName, ActorCallType_1.ActorCallType.REMINDER_METHOD);
    }
}
exports.default = ActorMethodContext;
