// @dapr/dapr/actors/runtime/ActorTimerData.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BufferSerializer_1 = __importDefault(require("./BufferSerializer"));
/**
 * Contains the actor reminder data
 */
class ActorTimerData {
    /**
     * @param reminderName the name of the actor reminder
     * @param state the state data passed to receiveReminder callback
     * @param callback the callback method to call
     */
    constructor(timerName, callback, state) {
        this.timerName = timerName;
        this.callback = callback;
        this.state = state;
    }
    getTimerName() {
        return this.timerName;
    }
    getState() {
        return this.state;
    }
    getCallback() {
        return this.callback;
    }
    /**
     * Return this class as an object
     */
    toObject() {
        return {
            callback: this.callback,
            data: this.state,
        };
    }
    static fromObject(reminderName, obj) {
        const serializer = new BufferSerializer_1.default();
        const callback = obj?.callback;
        const data = obj?.data;
        const deserializedData = serializer.deserialize(data);
        return new ActorTimerData(reminderName, callback, deserializedData);
    }
}
exports.default = ActorTimerData;
