// @dapr/dapr/logger/ConsoleLoggerService.js
"use strict";
/*
Copyright 2022 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleLoggerService = void 0;
/**
 * ConsoleLoggerService uses the in-built console module to log messages.
 */
class ConsoleLoggerService {
    error(message, ...optionalParams) {
        console.error(`${this.getTime()} ERROR ${message}`, ...optionalParams);
    }
    warn(message, ...optionalParams) {
        console.warn(`${this.getTime()} WARN ${message}`, ...optionalParams);
    }
    info(message, ...optionalParams) {
        console.info(`${this.getTime()} INFO ${message}`, ...optionalParams);
    }
    verbose(message, ...optionalParams) {
        console.log(`${this.getTime()} VERBOSE ${message}`, ...optionalParams);
    }
    debug(message, ...optionalParams) {
        console.debug(`${this.getTime()} DEBUG ${message}`, ...optionalParams);
    }
    getTime() {
        return new Date().toISOString();
    }
}
exports.ConsoleLoggerService = ConsoleLoggerService;
