// @dapr/dapr/workflow/runtime/WorkflowRuntime.js
"use strict";
/*
Copyright 2024 The Dapr Authors
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const durabletask_js_1 = require("@microsoft/durabletask-js");
const WorkflowActivityContext_1 = __importDefault(require("./WorkflowActivityContext"));
const WorkflowContext_1 = __importDefault(require("./WorkflowContext"));
const index_1 = require("../internal/index");
const internal_1 = require("../internal");
/**
 * Contains methods to register workflows and activities.
 */
class WorkflowRuntime {
    /**
     * Initializes a new instance of the WorkflowRuntime.
     * @param {WorkflowClientOptions | undefined} options - Additional options for configuring WorkflowRuntime.
     */
    constructor(options = {}) {
        const grpcEndpoint = (0, index_1.generateEndpoint)(options);
        options.daprApiToken = (0, index_1.getDaprApiToken)(options);
        this.worker = this.buildInnerWorker(grpcEndpoint, options);
    }
    buildInnerWorker(grpcEndpoint, options) {
        let innerOptions = options?.grpcOptions;
        if (options.daprApiToken !== undefined && options.daprApiToken !== "") {
            innerOptions = {
                ...innerOptions,
                interceptors: [(0, index_1.generateApiTokenClientInterceptors)(options), ...(innerOptions?.interceptors ?? [])],
            };
        }
        return new durabletask_js_1.TaskHubGrpcWorker(grpcEndpoint.endpoint, innerOptions, grpcEndpoint.tls);
    }
    /**
     * Registers a Workflow implementation for handling orchestrations.
     *
     * @param {TWorkflow} workflow - The instance of the Workflow class being registered.
     */
    registerWorkflow(workflow) {
        const name = (0, internal_1.getFunctionName)(workflow);
        const workflowWrapper = (ctx, input) => {
            const workflowContext = new WorkflowContext_1.default(ctx);
            return workflow(workflowContext, input);
        };
        this.worker.addNamedOrchestrator(name, workflowWrapper);
        return this;
    }
    /**
     * Registers a Workflow implementation for handling orchestrations with a given name.
     * The name provided need not be same as workflow name.
     *
     * @param {string} name - The name or identifier for the registered Workflow.
     * @param {TWorkflow} workflow - The instance of the Workflow class being registered.
     */
    registerWorkflowWithName(name, workflow) {
        const workflowWrapper = (ctx, input) => {
            const workflowContext = new WorkflowContext_1.default(ctx);
            return workflow(workflowContext, input);
        };
        this.worker.addNamedOrchestrator(name, workflowWrapper);
        return this;
    }
    /**
     * Registers an Activity object.
     *
     * @param {TWorkflowActivity<TInput, TOutput>} fn - The instance of the WorkflowActivity class being registered.
     * @returns {WorkflowRuntime} The current instance of WorkflowRuntime.
     */
    registerActivity(fn) {
        const name = (0, internal_1.getFunctionName)(fn);
        const activityWrapper = (ctx, intput) => {
            const wfActivityContext = new WorkflowActivityContext_1.default(ctx);
            return fn(wfActivityContext, intput);
        };
        this.worker.addNamedActivity(name, activityWrapper);
        return this;
    }
    /**
     * Registers an Activity object with a given name.
     * The name provided need not be same as WorkflowActivity name.
     *
     * @param {string} name - The name or identifier for the registered Activity.
     * @param {TWorkflowActivity<TInput, TOutput>} fn - The instance of the WorkflowActivity class being registered.
     * @returns {WorkflowRuntime} The current instance of WorkflowRuntime.
     */
    registerActivityWithName(name, fn) {
        const activityWrapper = (ctx, intput) => {
            const wfActivityContext = new WorkflowActivityContext_1.default(ctx);
            return fn(wfActivityContext, intput);
        };
        this.worker.addNamedActivity(name, activityWrapper);
        return this;
    }
    /**
     * Start the Workflow runtime processing items and block.
     */
    async start() {
        await this.worker.start();
    }
    /**
     * Stop the worker and wait for any pending work items to complete
     */
    async stop() {
        await this.worker.stop();
    }
}
exports.default = WorkflowRuntime;
